olc_art.hpp

```diff
@@ -264,6 +759,72 @@ class olc_db final {
   friend struct detail::olc_impl_helpers;
 };
 
+///
+/// ART iterator implementation.
+///
+
+inline std::optional<const unodb::key> olc_db::iterator::get_key() noexcept {
+  // Note: If the iterator is on a leaf, we return the key for that
+  // leaf regardless of whether the leaf has been deleted.  This is
+  // part of the design semantics for the OLC ART scan.
+  //
+  // TODO(thompsonbry) Eventually this will need to use the stack to
+  // reconstruct the key from the path from the root to this leaf.
+  // Right now it is relying on the fact that simple fixed width keys
+  // are stored directly in the leaves.
+  if (!valid()) return {};  // not positioned on anything.
+  const auto& e = stack_.top();
+  const auto& node = e.node;
+  UNODB_DETAIL_ASSERT(node.type() == node_type::LEAF);     // On a leaf.
+  const auto* const aleaf{node.ptr<detail::olc_leaf*>()};  // current leaf.
+  key_ =
+      aleaf->get_key().decode();  // decode the key into the iterator's buffer
+  return key_;                    // return pointer to the internal key buffer.
+}
+
+inline std::optional<const value_view> olc_db::iterator::get_val()
+    const noexcept {
+  // Note: If the iterator is on a leaf, we return the value for
+  // that leaf regardless of whether the leaf has been deleted.
+  // This is part of the design semantics for the OLC ART scan.
+  //
+  if (!valid()) return {};  // not positioned on anything.
+  const auto& e = stack_.top();
+  const auto& node = e.node;
+  UNODB_DETAIL_ASSERT(node.type() == node_type::LEAF);     // On a leaf.
+  const auto* const aleaf{node.ptr<detail::olc_leaf*>()};  // current leaf.
+  return aleaf->get_value_view();
+}
+
+inline bool olc_db::iterator::operator==(const iterator& other) const noexcept {
+  if (&db_ != &other.db_) return false;  // different tree?
+  if (stack_.empty() != other.stack_.empty())
+    return false;                   // one stack is empty and the other is not?
+  if (stack_.empty()) return true;  // both empty.
```

Comment:
In the case of OLC, does it make sense to compare stack lengths too?

Reply:
The check is always against an empty other iterator (the end() iterator).  The iterator class is not exposed and the scan is the only place where could be checked.  In fact, I believe that the end() comparison is actually only used from the test suite and scan() runs while valid() is true or (if the to_key is specified) until GTE the to_key is achieved.

Hence, not worth optimizing.  Mainly there for the contract.

Reply:
I see. So on one hand the check `foo == bar.end()` is idiomatic, on the other, it's all dead (and untested) code past the emptiness check. Not optimal, but probably best left alone